### ISSUE TITLE
Kyyti: Vanhojen myöhästymismaksujen date-kentän korjaus ja maksujen mitätöinti

### ISSUE THREAD

**Issue by Tuomas-Kyyti** [KohaSuomi/Koha#1036]
### Tukipyyntö

Kyytin myöhästymismaksut ja lainakohtaiset maksimimaksut nousivat syyskuun alussa 2023, jolloin kaikkien lainassa olleiden niteiden maksurivit myös päivittyivät. Koska date-kenttä päivittyi maksujen päivityksessä, maksut eivät tulleet 1.1.2024 maksujen vanhentamisajoon mukaan.

Näiden maksujen päivämäärät pitää korjata lähemmäs alkuperäistä maksujen luontipäivämäärää. Tämän jälkeen maksut mitätöidään writeoffauksella ja maksujen note-kenttään tieto mitätöinnistä.

Kaikissa näissä tapauksissa maksurivi on päivittynyt 2.9.2023, kellonaika heittää muutamilla minuuteilla.

**Comment by lmstrand** [comment]
Kyytissä on kolmella tapaa description-kenttään merkkauksia maksuriveillä, joilla on tuo 2.9.2023 päivämääräksi muuttunut. Jokaiselle pitää tehdä oma sql-rimpsunsa, jotta saa päivämäärän kiinni (suurimmalle osalle toimii outissa ja vaskissakin käytetty sql-lauseke jos päivämäärän/kellonajan perässä ei ole mitään).

Hits 2017 / 05.03.2020 23:59
Sokkoleikki / 31.08.2023 286N0N980873 
Vappu joka päivä : 30.08.2023 28622102800523

**Comment by atarnus** [comment]
Tällä voisi kokeilla, tarttuuhan oikeat rivit matkaan:

```
SELECT i.date_due + INTERVAL 29 DAY + INTERVAL 1 SECOND AS pvm, al.*
  FROM accountlines al
       INNER JOIN issues i USING(issue_id)
 WHERE al.debit_type_code IN ('OVERDUE')
   AND DATE(al.date) = '2023-09-02'
   AND al.amountoutstanding > 0
   AND i.date_due + 35 < '2023-09-02'
 ORDER BY i.date_due DESC;
```

Voidaan hioa, ja kun on oikein, niin ajaa sitten nuo uudet päiväykset, ja sama sitten old_issues-taulun kanssa.

**Comment by Tuomas-Kyyti** [comment]
Ei tule tuloksia tällä raportilla.

**Comment by atarnus** [comment]
Toimiskohan nyt kun tuo eräpäivän tarkistuskin laskee siihen lisää päiviä eikä vain numeroita? :)

**Comment by Tuomas-Kyyti** [comment]
Nyt tulee tuloksia, mutta niiden ei kuulukaan vanhentua, kun ovat viime vuonna syntyneitä maksuja. Mutta jos rivin 
AND i.date_due + INTERVAL 35 DAY < '2023-09-02' muuttaa muotoon 
AND i.date_due < '2019-01-01',
niin alkaa näyttää oikealta.

**Comment by atarnus** [comment]
Kyllä, tuolla muokkauksella saa kiinni ne, joiden kuuluisi vanhentua _nyt._ Tarkoitus olisi kuitenkin saada se pvm korjattua myös sellaisille maksuille, jotka vanhenevat tulevina vuosina, että ei tarvitsisi sitten taas ensi vuoden alussa ottaa uutta satsia. Mutta voidaan rajata ne vaikka ennen vuotta 2023 niin ei nyt ainakaan mitään ihan tuoreita nappaannu mukaan.

Sitten kun nuo päiväykset on korjattu, niin annetaan niiden vanhentua, joille tulee se päiväys ennen 2019, ja ensi vuonna sitten ne 2019 päivätyt jne. Mutta minä voin testailla tässä nyt itse tarkemmin. Tiedotetaan sitten miten jatketaan.

**Comment by atarnus** [comment]
Näköjään piti vaihtaa lisätty sekunti minuutiksi myös. Näille riveille ajaisin nyt uuden pvm:n.

```
1035 riviä (issues):

SELECT i.date_due + INTERVAL 29 DAY + INTERVAL 1 MINUTE AS pvm, al.*
  FROM accountlines al
       INNER JOIN issues i USING(issue_id)
 WHERE al.debit_type_code IN ('OVERDUE')
   AND DATE(al.date) = '2023-09-02'
   AND al.amountoutstanding > 0
   AND i.date_due < '2023-01-01'
 ORDER BY i.date_due DESC;


59 riviä (old issues):

SELECT i.date_due + INTERVAL 29 DAY + INTERVAL 1 MINUTE AS pvm, al.*
  FROM accountlines al
       INNER JOIN old_issues i USING(issue_id)
 WHERE al.debit_type_code IN ('OVERDUE')
   AND DATE(al.date) = '2023-09-02'
   AND al.amountoutstanding > 0
   AND i.date_due < '2023-01-01'
 ORDER BY i.date_due DESC;
```

**Comment by lmstrand** [comment]
Kyytiin ajettu uudet päivämäärät:
```
MariaDB [kyytiprod]> update accountlines set date = STR_TO_DATE(CONCAT(REGEXP_SUBSTR(description, '[0-9]{2}.[0-9]{2}.[0-9]{4}'), ' 00:00:00'), '%d.%m.%Y %H:%i:%s') + INTERVAL 30 DAY
    -> where accountlines_id in
    -> (
    -> select accountlines_id FROM accountlines al where date(date) = '2023-09-02' and debit_type_code = 'OVERDUE'
    -> );
Query OK, 1742 rows affected (17.657 sec)
Rows matched: 1742  Changed: 1742  Warnings: 0
```

**Comment by lmstrand** [comment]
Ajettu maksujen mitätöinti 5v ajolla uusiksi. 369 OVERDUE-riviä maksuja mitätöityi.

**Comment by Tuomas-Kyyti** [comment]
Tämä lienee kunnossa.
